bundles/org.openhab.voice.googletts/src/main/java/org/openhab/voice/googletts/internal/GoogleCloudAPI.java

```diff
@@ -0,0 +1,469 @@
+    private final Logger logger = LoggerFactory.getLogger(GoogleCloudAPI.class);
+
+    /**
+     * Supported voices and locales
+     */
+    private final Map<Locale, Set<GoogleTTSVoice>> voices = new HashMap<>();
+
+    /**
+     * Cache folder
+     */
+    private File cacheFolder;
+
+    /**
+     * Configuration
+     */
+    private @Nullable GoogleTTSConfig config;
+
+    /**
+     * Status flag
+     */
+    private boolean initialized;
+
+    private final Gson gson = new GsonBuilder().create();
+    private final ConfigurationAdmin configAdmin;
+    private final OAuthFactory oAuthFactory;
+
+    private @Nullable OAuthClientService oAuthService;
+
+    /**
+     * Constructor.
+     *
+     * @param cacheFolder Service cache folder
+     */
+    GoogleCloudAPI(ConfigurationAdmin configAdmin, OAuthFactory oAuthFactory, File cacheFolder) {
+        this.configAdmin = configAdmin;
+        this.oAuthFactory = oAuthFactory;
+        this.cacheFolder = cacheFolder;
+    }
+
+    /**
+     * Configuration update.
+     *
+     * @param config New configuration.
+     */
+    void setConfig(GoogleTTSConfig config) {
+        this.config = config;
+
+        String clientId = config.clientId;
+        String clientSecret = config.clientSecret;
+        if (clientId != null && !clientId.isEmpty() && clientSecret != null && !clientSecret.isEmpty()) {
+            try {
+                final OAuthClientService localOAuthService = oAuthFactory.createOAuthClientService(
+                        GoogleTTSService.SERVICE_PID, GCP_TOKEN_URI, GCP_AUTH_URI, clientId, clientSecret, GCP_SCOPE,
+                        false);
+                oAuthService = localOAuthService;
+
+                if (checkOnAuthCode()) {
+                    initialized = true;
+                    initVoices();
+                }
+            } catch (AuthenticationException | IOException ex) {
+                logger.warn("Error initializing Google Cloud TTS service: {}", ex.getMessage());
+                oAuthService = null;
+                initialized = false;
+            }
+        } else {
+            oAuthService = null;
+            initialized = false;
+            voices.clear();
+        }
+
+        // maintain cache
+        if (config.purgeCache != null && config.purgeCache) {
+            File[] files = cacheFolder.listFiles();
+            if (files != null && files.length > 0) {
+                Arrays.stream(files).forEach(File::delete);
+            }
+            logger.debug("Cache purged.");
+        }
+    }
+
+    /**
+     * Fetches the OAuth2 tokens from Google Cloud Platform if the auth-code is set in the configuration. If successful
+     * the auth-code will be removed from the configuration.
+     *
+     * @return true if the auth-code was not set or if the auth-code was successfully used to get a new refresh and
+     *         access token
+     */
+    private boolean checkOnAuthCode() throws AuthenticationException, IOException {
```

Comment:
Since it will either throw exceptions or return `true` the result isn't really used since it will never be `false`. So it would make sense to turn this into a `void` method with a name like `refreshAccessToken()`.

Reply:
Done. I did not think about that during development.